x/leverage/keeper/keeper.go

```diff
@@ -146,11 +146,25 @@ func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken
 
 	if amountFromCollateral.IsPositive() {
 		if k.GetCollateralSetting(ctx, lenderAddr, uToken.Denom) {
-			// TODO: Calculate lender's borrow limit and current borrowed value, if any.
-			// Prevent withdrawing collateral when it would bring user borrow limit below
-			// current borrowed value.
-			//
-			// ref: https://github.com/umee-network/umee/issues/213
+			// Calculate current borrowed value
+			borrowed := k.GetBorrowerBorrows(ctx, lenderAddr)
+			borrowedValue, err := k.TotalTokenValue(ctx, borrowed)
+			if err != nil {
+				return err
+			}
+
+			// Calculate what borrow limit will be AFTER this withdrawal
+			collateral := k.GetBorrowerCollateral(ctx, lenderAddr)
+			collateralToWithdraw := sdk.NewCoins(sdk.NewCoin(uToken.Denom, amountFromCollateral))
+			newBorrowLimit, err := k.CalculateBorrowLimit(ctx, collateral.Sub(collateralToWithdraw))
+			if err != nil {
+				return err
+			}
+
+			// Return error if borrow limit would drop below borrowed value
+			if borrowedValue.GT(newBorrowLimit) {
+				return sdkerrors.Wrap(types.ErrBorrowLimitLow, newBorrowLimit.String())
+			}
```

Comment:
The main reason for this change is to prohibit withdrawing loaned assets that if withdrawn, would put the account below the liquidation ration?

I'm not quite following why check borrowed value when withdrawing an asset. Is this what Compound does?

Reply:
They prevent withdrawals (and regular transfers) that would reduce the borrow limit below current borrowed value, yes.